docs/design/2021-02-09-ttl-table.md

```diff
@@ -0,0 +1,45 @@
+# Proposal: Time to live (TTL) Table
+
+- Author(s):     [Xiaoguang Sun](https://github.com/sunxiaoguang)
+- Last updated:  2021-02-09
+- Discussion at: https://github.com/pingcap/tidb/issues/22762
+
+## Abstract
+TTL Table makes it possible to automatically garbage collect data by user defined retention policy at either fine or coarse granularity.
+
+## Background
+There are application scenarios that data is only valuable for a certain period of time after ingestion and can be deleted permanently after expiration. Tracing, audit logs or push notification with expiration are examples of such applications. With the help of TTL tables, users can be relieved from taking care of the data life cycle in such tables. Therefore people are more willing to use TiDB as a general storage for such scenarios.
+
+## Proposal
+Introduce new `TTL` and `TTL_GRANULARITY` table options when creating or altering tables. By specifying `TTL`, users can expect expired data to be removed automatically. Additionally, users can also choose from `ROW` or `PARTITION` for `TTL_GRANULARITY` option to trade off collect granularity and cost to run such garbage collection. `ROW` mode evaluates expiry time against each row of data and reclaim space at row basis. This gives us the finest granularity and allows the most accurate expiration timing. `PARTITION` mode on the other hand partitions data according to its last update time. A background timer is going to rollover partitions and truncate all expired data within the oldest partition all at once.
```

Comment:
How is the expiry time calculated? IMO, it's `current timestamp - update timestamp` rather than `current timestamp - insert timestamp`, right?
What if an update statement only updates the row record but not the index, or it only updates some indexes but not other indexes? Some of the indexes will be collected earlier, which breaks the snapshot isolation.

Reply:
Yes you are right, we didn't consider the case where only data get updated and it breaks many things. So in order to fix this, maybe we should update both data and index even for the case when it is not necessary to do so for none TTL table.